pdg/src/info.rs

```diff
@@ -67,18 +67,597 @@ fn set_flow_info(g: &mut Graph) {
+        let c1 = mk_copy(&mut g, b1);
+        // *c = 0;
+        let c2 = mk_store_addr(&mut g, c1);
+        // *c = 0;
+        let b3 = mk_store_addr(&mut g, b1);
+
+        let pdg = build_pdg(g);
+        assert!(info(&pdg, a).unique);
+        assert!(info(&pdg, b1).unique);
+        assert!(info(&pdg, b2).unique);
+        assert!(info(&pdg, b3).unique);
+        assert!(info(&pdg, c1).unique);
+        assert!(info(&pdg, c2).unique);
+    }
+
+    /// ```rust
+    /// let mut a = 0;
+    /// let b = &mut a;
+    /// *b = 0;
+    /// let c = &mut *b;
+    /// *c = 0;
+    /// *b = 0;
+    /// *c = 0;
+    /// ```
+    ///
+    /// ```text
+    /// A
+    /// |
+    /// B1
+    /// +-B2
+    /// +----C1
+    /// |    +-C2
+    /// B3   |
+    ///      C3
+    /// ```
+    #[test]
+    fn unique_sub_borrow_bad() {
+        let mut g = Graph::default();
+
+        // let mut a = 0;
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        // let b = &mut a;
+        let b1 = mk_copy(&mut g, a);
+        // *b = 0;
+        let b2 = mk_store_addr(&mut g, b1);
+        // let c = &mut *b;
+        let c1 = mk_copy(&mut g, b1);
+        // *c = 0;
+        let c2 = mk_store_addr(&mut g, c1);
+        // *b = 0;
+        let b3 = mk_store_addr(&mut g, b1);
+        // *c = 0;
+        let c3 = mk_store_addr(&mut g, c1);
+
+        let pdg = build_pdg(g);
+        assert!(info(&pdg, a).unique);
+        assert!(!info(&pdg, b1).unique);
+        assert!(!info(&pdg, b2).unique);
+        assert!(!info(&pdg, b3).unique);
+        assert!(!info(&pdg, c1).unique);
+        assert!(!info(&pdg, c2).unique);
+        assert!(!info(&pdg, c3).unique);
+    }
+
+    /// ```rust
+    /// let mut a = Point { x: 0, y: 0 };
+    /// let b = &mut a.x;
+    /// let c = &mut a.y;
+    /// *b = 1;
+    /// *c = 2;
+    /// ```
+    ///
+    /// ```text
+    /// A
+    /// |-------
+    /// |x     |
+    /// B1     |y
+    /// |      C1
+    /// B2     |
+    ///        C2
+    /// ```
+    #[test]
+    fn okay_use_different_fields() {
+        let mut g = Graph::default();
+
+        // let mut a = Point { x: 0, y: 0 };
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        // let b = &mut a.x;
+        let b11 = mk_field(&mut g, a, 0_u32);
+        let b1 = mk_copy(&mut g, b11);
+        // let c = &mut a.y;
+        let c11 = mk_field(&mut g, a, 1_u32);
+        let c1 = mk_copy(&mut g, c11);
+        // *b = 1;
+        let b2 = mk_store_addr(&mut g, b1);
+        // *c = 2;
+        let c2 = mk_store_addr(&mut g, c1);
+
+        let pdg = build_pdg(g);
+        assert!(info(&pdg, a).unique);
+        assert!(info(&pdg, b1).unique);
+        assert!(info(&pdg, b2).unique);
+        assert!(info(&pdg, c1).unique);
+        assert!(info(&pdg, c2).unique);
+    }
+
+    /// ```rust
+    /// let mut a = Point { x: 0, y: 0 };
+    /// let j = &mut a;
+    /// let b = &mut j.x;
+    /// let c = &mut j.x;
+    /// *b = 1;
+    /// *c = 2;
+    /// *(a.y) = 3;
```

Comment:
There are similar issues on several other tests.  In general, accessing a field behind a reference should be either `(*a).y` or just `a.y`, rather than `*(a.y)` (which indicates dereferencing a reference stored in `a.y` instead).